server/src/main/scala/services/graphstages/CrunchLoadGraphStage.scala

```diff
@@ -0,0 +1,174 @@
+package services.graphstages
+
+import akka.stream._
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import drt.shared.CrunchApi.{CrunchMinute, CrunchMinutes, MillisSinceEpoch}
+import drt.shared.FlightsApi.{QueueName, TerminalName}
+import drt.shared._
+import org.slf4j.{Logger, LoggerFactory}
+import services.graphstages.Crunch._
+import services.{OptimizerConfig, OptimizerCrunchResult, SDate}
+
+import scala.collection.immutable.Map
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+
+
+class CrunchLoadGraphStage(optionalInitialCrunchMinutes: Option[CrunchMinutes],
+                           airportConfig: AirportConfig,
+                           expireAfterMillis: MillisSinceEpoch,
+                           now: () => SDateLike,
+                           crunch: (Seq[Double], Seq[Int], Seq[Int], OptimizerConfig) => Try[OptimizerCrunchResult])
+  extends GraphStage[FlowShape[Loads, CrunchMinutes]] {
+
+  val inLoads: Inlet[Loads] = Inlet[Loads]("inLoads.in")
+  val outCrunchMinutes: Outlet[CrunchMinutes] = Outlet[CrunchMinutes]("outCrunchMinutes.out")
+
+  override val shape = new FlowShape(inLoads, outCrunchMinutes)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    var loadMinutes: Map[Int, LoadMinute] = Map()
+    var existingCrunchMinutes: Map[Int, CrunchMinute] = Map()
+    var crunchMinutesToPush: Map[Int, CrunchMinute] = Map()
+
+    val log: Logger = LoggerFactory.getLogger(getClass)
+
+    override def preStart(): Unit = {
+
+      //      loadMinutes = optionalInitialLoads match {
+      //        case Some(Loads(lm)) =>
+      //          log.info(s"Received ${lm.size} initial loads")
+      //          purgeExpiredLoads(lm)
+      //        case _ =>
+      //          log.warn(s"Did not receive any loads to initialise with")
+      //          Set()
+      //      }
+
+      super.preStart()
+    }
+
+    setHandler(inLoads, new InHandler {
+      override def onPush(): Unit = {
+        val incomingLoads = grab(inLoads)
+        log.info(s"Received ${incomingLoads.loadMinutes.size} loads")
+        /*
+        1) establish crunch window - start minute & end minute from incoming set of loads
```

Comment:
I'm weary of this kind of comment because they tend to get outdated quickly, I think the code below is probably clear enough that you don't need it. Or possibly you could break each block into a more clearly named function.

Reply:
sorry yes that wasn't meant to be left in. it was just for me to reference when building the thing